sdk/storage/azure-storage-files-shares/src/share_client.cpp

```diff
@@ -41,31 +41,44 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
       const ShareClientOptions& options)
       : m_shareUrl(shareUrl)
   {
-    Azure::Core::Http::TelemetryOptions telemetryPolicyOptions;
-    telemetryPolicyOptions.ApplicationId = options.ApplicationId;
+    ShareClientOptions newOptions = options;
```

Comment:
There's a bug at this line and I need some suggestion on how to fix it properly.

`ShareClientOptions` inherits from `ClientOptions` and it has an extra field of `ApiVersion`.
When we make a copy of this `ShareClientOptions`, the `ApiVersion` string is copied to new options.
Every field of `ClientOptions` is copied by the copy constructor of `ClientOptions`.
However, [copy constructor](https://github.com/Azure/azure-sdk-for-cpp/blob/a475980de62874a61c42f0aa5b64a65742e497ac/sdk/core/azure-core/inc/azure/core/internal/client_options.hpp#L54) of `ClientOptions` only copies policies and leaves out all other fields. So other fields are not copied, and this is the bug.

@vhvb1989 Please suggest

Reply:
Why do you need to copy them?  Is it for copying the ShareClientOptions within the service (like from the clientMethod to the protocol rest method? can you use const& for that? 

When the ShareClientOptions is passed to the HttpClient, any extra field in the options doesn't care for the HttpPipeline.
For instance, we would create a ValuePolicy which contains the `ApiVersion` as an option inside the policy, then we add that policy to the per-call-service policies array and send that array to the httpPipeline constructor, we don't need to persist the extended options from the `shareClientOptions`.

Reply:
> Why do you need to copy them? Is it for copying the ShareClientOptions within the service (like from the clientMethod to the protocol rest method? can you use const& for that?

You can stop me from copying options. But this options is exposed to customers, you cannot stop them

Reply:
because user-customized per-retry policies come after service per-retry policies, if I want to make key auth policy the last one, I have to make a copy and append it into user-customized per-retry policies.

Reply:
We need @CaseyCarter  here

Reply:
We also need @JeffreyRichter here

Reply:
oh wait, don't we only need to update the copy and move ctor from clientOptions here?

Reply:
@Jinming-Hu  here: https://github.com/Azure/azure-sdk-for-cpp/pull/1788
That should solve this issue you are getting.